### PR TITLE
[INFRA] Introduce `mvnd`  to speed up CI jobs of Dependency, Licence and Style Check

### PR DESCRIPTION
**Pull request by bowenliang123**
#### Motivation:
- speed up maven building and testing with `mvnd` (https://github.com/apache/maven-mvnd). `mvnd` itself embeds exactly the same distribution of maven with more improvements and features on the client side and compilation server in daemon mode.
- also inspired by custom github action for mvnd installing in Apache Camel (https://github.com/apache/camel/blob/main/.github/actions/install-mvnd/action.yml)

#### Changes in this PR:
- introducing `mvnd` by adding `build/mvnd` script
- use `mvnd` version `0.9.0` which embeds  `maven` 3.8.7. The maven version embedded in `mvnd` is also check and guaranteed  the same as maven version required in pom by `build/mvnd` script.
- use mvnd in CI jobs of Depedency Check and Style Check

#### Comparision
- CI jobs
(both with maven dependencies cached in local repo)

Job | with `build/mvn` | with `build/mvnd` 
--- | --- | --- 
Dependency Check | 9min1sec;[see log](https://github.com/apache/kyuubi/actions/runs/3966175262/jobs/6796688187) | 6min 46 sec ; [see log](https://github.com/apache/kyuubi/actions/runs/4124518481/jobs/7123867015) 
Style Check | 10min32sec ; [see log](https://github.com/apache/kyuubi/actions/runs/4122613274/jobs/7119603428)  | 7min50s ; [see log](https://github.com/apache/kyuubi/actions/runs/4125013599/jobs/7125006946)
Licence Check | 32s ; [see log](https://github.com/apache/kyuubi/actions/runs/4130799867/jobs/7137835062) | 21s ; [see log](https://github.com/apache/kyuubi/actions/runs/4130923126/jobs/7138088114)

- building entire maven project on local machine (skipping test running and style checking with `fast` profile)

`build/mvnd clean install -Pfast`
Pasted Graphic

`build/mvn clean install -Pfast`
Pasted Graphic 1

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: build/mvnd [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      x86_64*)    arch=amd64;;
+      arm64*)     arch=aarch64;;
+      *)          arch="UNKNOWN:${unameOsOut}"
+  esac
+  echo "$arch"
+}
+
+# Determine the Maven version from the root pom.xml file and
+# install maven under the build/ folder if needed.
+install_mvn() {
```
Comment (yaooqinn) [inline build/mvnd:96]: install_mvnd

File: build/mvnd [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Determine the current working directory
+_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+# Preserve the calling directory
+_CALLING_DIR="$(pwd)"
+# Options used during compilation
+_COMPILE_JVM_OPTS="-Xms2g -Xmx2g -XX:ReservedCodeCacheSize=1g -Xss128m"
```
Comment (cfmcgrady) [inline build/mvnd:25]: Github Action limit memory: 6947m, we may increase the max heap size?

File: .github/workflows/dep.yml [hunk 3]
```diff
@@ -23,8 +23,11 @@ on:
       - master
       - branch-*
     paths:
-      # dependency check happens only pom changes
+      # when pom changes
       - '**/pom.xml'
+      # when dependency workflow changes
+      - .github/workflows/dep.yaml
```
Comment (cfmcgrady) [inline .github/workflows/dep.yml:29]: unnecessary changes?

**Review by cfmcgrady — APPROVED** [review]

File: build/mvnd (hunk 2 above)
Comment (bowenliang123) [reply build/mvnd:25]: `build/mvnd` is designed for usage of both on local machine and also the github action, so I would prefer keep -Xmx here the same as in `build/mvn`. And we could increase the memory limit in CI jobs by passing opts.

File: .github/workflows/dep.yml (hunk 3 above)
Comment (bowenliang123) [reply .github/workflows/dep.yml:29]: To verify using mvnd in dependency workflows. Changes in `dep.yaml` should trigger the jobs defined from the file itself.

**Comment by pan3793** [comment]
I suppose this target to 1.8?

**Comment by bowenliang123** [comment]
> I suppose this target to 1.8?
I think `mvnd` works fine with Java 8 or up. And there's also an mvnd specific option ` -Djava.home=` to set Java home for starting the daemon.
```
$ mvnd -h
...
mvnd specific options:
...
 -Djava.home=<path>                     Java home for starting the daemon
                                        Env. variable: JAVA_HOME
...
```

**Comment by pan3793** [comment]
Sorry for confusing, I mean target to Kyuubi 1.8.0

**Comment by bowenliang123** [comment]
> Sorry for confusing, I mean target to Kyuubi 1.8.0

This PR brings no changes in the source code and it is not part of the distribution. I think it's good to merging it to master/1.7 .

File: build/mvnd (hunk 2 above)
Comment (bowenliang123) [reply build/mvnd:25]: https://github.com/apache/spark/blob/master/.github/workflows/build_and_test.yml#L950
Marking the memory limit in Spark as above.

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master for 1.8

**Comment by aaime** [comment]
Hi, I'm considering using mvnd in github actions too. I see mvnd usage was eventually reverted in [this commit](https://github.com/apache/kyuubi/commit/6431225b441193647c6134ab2587bd4f96d87638), and I'm wondering what was the reason? I would like to find out about drawbacks of mvnd adoption in github actions before I have to learn the hard way :rofl:
